NotificationHubs/src/com/windowsazure/messaging/Notification.java

```diff
@@ -6,245 +10,381 @@
 import org.apache.http.entity.ContentType;
 
 /**
- * 
  * Class representing a generic notification.
- * 
  */
 public class Notification {
 
-	private Map<String, String> headers = new HashMap<String, String>();
-	private String body;
-	private ContentType contentType;
-
-	/**
-	 * Utility method to set up a native notification for WNS. Sets the
-	 * X-WNS-Type headers based on the body provided. If you want to send raw
-	 * notifications you have to set the X-WNS header and ContentType after creating this
-	 * notification or use createWindowsRawNotification method
-	 * 
-	 * @param body
-	 * @return
-	 */
-	public static Notification createWindowsNotification(String body) {
-		Notification n = new Notification();
-		n.body = body;
-
-		n.headers.put("ServiceBusNotification-Format", "windows");
-
-		if (body.contains("<toast>"))
-			n.headers.put("X-WNS-Type", "wns/toast");
-		if (body.contains("<tile>"))
-			n.headers.put("X-WNS-Type", "wns/tile");
-		if (body.contains("<badge>"))
-			n.headers.put("X-WNS-Type", "wns/badge");
-
-		if (body.startsWith("<")) {
-			n.contentType = ContentType.APPLICATION_XML;
-		}
-
-		return n;
-	}
-	
-	/**
-	 * Utility method to set up a native notification for WNS. Sets the
-	 * X-WNS-Type header to "wns/raw" in order of sending of raw notification.
-	 * 
-	 * @param body
-	 * @return
-	 */
-	public static Notification createWindowsRawNotification(String body) {
-		Notification n = new Notification();
-		n.body = body;
-		n.headers.put("ServiceBusNotification-Format", "windows");
-		n.headers.put("X-WNS-Type", "wns/raw");
-		n.contentType = ContentType.APPLICATION_OCTET_STREAM;
-		return n;
-	}
-
-    /**
-     * Utility method to set up a native notification for APNs.
-     * An expiry Date of 1 day is set by default.
-     * @param body
-     * @return
+    Map<String, String> headers = new HashMap<>();
+    String body;
+    ContentType contentType;
+
+    /**
+     * Utility method to set up a native notification for WNS. Sets the X-WNS-Type
+     * headers based on the body provided. If you want to send raw notifications you
+     * have to set the X-WNS header and ContentType after creating this notification
+     * or use createWindowsRawNotification method
+     *
+     * @param body The body for the notification
+     * @return a new Windows Notification.
+     */
+    public static Notification createWindowsNotification(String body) {
+        Notification n = new Notification();
+        n.body = body;
+
+        n.headers.put("ServiceBusNotification-Format", "windows");
+
+        if (body.matches("^<toast[\\s\\S]*>[\\s\\S]+</toast>$"))
+            n.headers.put("X-WNS-Type", "wns/toast");
+        if (body.matches("^<tile[\\s\\S]*>[\\s\\S]+</tile>$"))
+            n.headers.put("X-WNS-Type", "wns/tile");
+        if (body.matches("^<badge[\\s\\S]*>[\\s\\S]+</badge>$"))
```

Comment:
So we previously only looked for the opening tag, but would miss if it included any attributes. This now captures the entire body. Do we have corresponding tests for this regexp? Seems easy to get wrong.

Reply:
For instance, in the case there were multiple "tile" instances in the body, the greedy match in the beginning would swallow all instances and turn it into one match.

Reply:
This didn't change from the previous release, only the spacing changed